repondeur/zam_repondeur/resources.py

```diff
@@ -0,0 +1,215 @@
+from typing import Iterator, List, Optional, cast
+
+from pyramid.request import Request
+
+from zam_repondeur.models import (
+    Amendement as AmendementModel,
+    Lecture as LectureModel,
+    DBSession,
+)
+
+
+class ResourceNotFound(Exception):
+    pass
+
+
+class Resource(dict):
+    """
+    Location-aware resource
+
+    See: https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/resources.html#location-aware-resources  # noqa
+    """
+
+    __name__: Optional[str] = None
+    __parent__: Optional["Resource"] = None
+
+    @property
+    def breadcrumbs_title(self) -> str:
+        raise NotImplementedError  # pragma: no cover
+
+    breadcrumbs_link: bool = True
+
+    def __init__(self, name: str, parent: "Resource") -> None:
+        self.__name__ = name
+        self.__parent__ = parent
+
+    @property
+    def parent(self) -> Optional["Resource"]:
+        return self.__parent__
+
+    @property
+    def parents(self) -> Iterator["Resource"]:
+        parent = self.parent
+        while parent is not None:
+            yield parent
+            parent = parent.parent
+
+    @property
+    def ancestors(self) -> List["Resource"]:
+        return list(reversed(list(self.parents)))
+
+    def add_child(self, child: "Resource") -> None:
+        self[child.__name__] = child
+
+
+class Root(Resource):
+    def __init__(self, request: Request) -> None:
+        self.add_child(LectureCollection(name="lectures", parent=self))
+
+
+class LectureCollection(Resource):
+
+    breadcrumbs_title = "Lectures"
+
+    def models(self) -> List[LectureModel]:
+        return LectureModel.all()
+
+    def __getitem__(self, key: str) -> Resource:
+        try:
+            chambre, session, num_texte, organe = key.split(".")
+        except ValueError:
+            raise KeyError
+        return LectureResource(
+            name=key,
+            parent=self,
+            chambre=chambre,
+            session=session,
+            num_texte=int(num_texte),
+            organe=organe,
+        )
+
+
+class LectureResource(Resource):
+    def __init__(
+        self,
+        name: str,
+        parent: Resource,
+        chambre: str,
+        session: str,
+        num_texte: int,
+        organe: str,
+    ) -> None:
+        super().__init__(name=name, parent=parent)
+        self.chambre = chambre
+        self.session = session
+        self.num_texte = num_texte
+        self.organe = organe
+        self.add_child(AmendementCollection(name="amendements", parent=self))
+        self.add_child(ArticleCollection(name="articles", parent=self))
+
+    def model(self) -> LectureModel:
+        lecture = LectureModel.get(
+            self.chambre, self.session, self.num_texte, self.organe
+        )
+        if lecture is None:
+            raise ResourceNotFound(self)
+        return lecture
+
+    @property
+    def breadcrumbs_title(self) -> str:
+        return str(self.model())
```

Comment:
Proposal: `return f"{self.model().dossier_legislatif} → {str(self.model())}"` otherwise we are loosing the information related to the text.

Reply:
Thanks, I missed that when rebasing!